File: dailyQuestion/2020/2020-07/07-28/python/solution_recurse.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def maxDepth(root: TreeNode) -> int:
    if not root :
        return 0
    return max(maxDepth(root.left), maxDepth(root.right)) + 1

if __name__ == "__main__" :
    node = TreeNode(3)
    node.left = TreeNode(9)
    node.right = TreeNode(20)
    node.right.left = TreeNode(15)
    node.right.right = TreeNode(7)
    result = maxDepth(node)
    print(result)

File: dailyQuestion/2020/2020-06/06-22/golang/solution_recurse.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	maxPath := math.MinInt32
	var maxGain func(root *TreeNode) int
	maxGain = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		leftGain := max(maxGain(root.Left), 0)
		rightGain := max(maxGain(root.Right), 0)
		maxPath = max(maxPath, root.Val+leftGain+rightGain)
		return root.Val + max(leftGain, rightGain)
	}
	maxGain(root)
	return maxPath
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {
	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}

	result := maxPathSum(root)

	fmt.Println(result)
}
